pkg/bootkube/create.go

```diff
@@ -43,143 +59,133 @@ func CreateAssets(config clientcmd.ClientConfig, manifestDir string, timeout tim
-// predict with regards to error handling.
-func createAssets(config clientcmd.ClientConfig, manifestDir string) error {
-	f := cmdutil.NewFactory(config)
-
-	shouldValidate := true
-	schema, err := f.Validator(shouldValidate, fmt.Sprintf("~/%s/%s", clientcmd.RecommendedHomeDir, clientcmd.RecommendedSchemaName))
+func apiTest(c clientcmd.ClientConfig) error {
+	config, err := c.ClientConfig()
 	if err != nil {
 		return err
 	}
-
-	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	client, err := kubernetes.NewForConfig(config)
 	if err != nil {
 		return err
 	}
 
-	mapper, _, err := f.UnstructuredObject()
-	if err != nil {
-		return err
+	// API Server is responding
+	healthStatus := 0
+	client.Discovery().RESTClient().Get().AbsPath("/healthz").Do().StatusCode(&healthStatus)
+	if healthStatus != http.StatusOK {
+		return fmt.Errorf("API Server http status: %d", healthStatus)
 	}
 
-	builder, err := f.NewUnstructuredBuilder(true)
-	if err != nil {
-		return err
+	// System namespace has been created
+	_, err = client.CoreV1().Namespaces().Get("kube-system", metav1.GetOptions{})
+	return err
+}
+
+type manifest struct {
+	kind       string
+	apiVersion string
+	namespace  string
+	name       string
+	raw        []byte
+
+	filepath string
+}
+
+func (m manifest) String() string {
+	if m.namespace == "" {
+		return fmt.Sprintf("%s %s %s", m.filepath, m.kind, m.name)
 	}
+	return fmt.Sprintf("%s %s %s/%s", m.filepath, m.kind, m.namespace, m.name)
+}
 
-	filenameOpts := &resource.FilenameOptions{
-		Filenames: []string{manifestDir},
-		Recursive: false,
+type creater struct {
+	client *rest.RESTClient
+
+	// mapper maps resource kinds ("ConfigMap") with their pluralized URL
+	// path ("configmaps") using the discovery APIs.
+	mapper *resourceMapper
+}
+
+func newCreater(c *rest.Config) (*creater, error) {
+	c.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: scheme.Codecs}
+	client, err := rest.UnversionedRESTClientFor(c)
+	if err != nil {
+		return nil, err
 	}
 
-	r := builder.
-		Schema(schema).
-		ContinueOnError().
-		NamespaceParam(cmdNamespace).DefaultNamespace().
-		FilenameParam(enforceNamespace, filenameOpts).
-		SelectorParam("").
-		Flatten().
-		Do()
-	err = r.Err()
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(c)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	count := 0
-	err = r.Visit(func(info *resource.Info, err error) error {
-		if err != nil {
-			return err
-		}
+	return &creater{
+		mapper: newResourceMapper(discoveryClient),
+		client: client,
+	}, nil
+}
 
-		obj, err := resource.NewHelper(info.Client, info.Mapping).Create(info.Namespace, true, info.Object)
-		if err != nil {
-			if apierrors.IsAlreadyExists(err) {
-				count++
-				return nil
-			}
-			return cmdutil.AddSourceToErr("creating", info.Source, err)
+func (c *creater) createManifests(manifests []manifest) error {
+	var crds, other []manifest
+	for _, m := range manifests {
```

Comment:
This is a bit of a change in behavior. Prior, if the resource it just created happened to be a CRD, it would then wait until the api resource was available (so we knew that creating subsequent instances of that CRD would be successful).

If we scan the list of manifests and create all CRDs upfront - that does remove some of the ability to ensure strict ordering of all asset creation (e.g. `01-foo, 02-foo`). Off the top of my head I don't know that this would be an issue - but I wonder if the minor speed improvement is worth removing that ability to order all object creation.

Reply:
The `manifest` type has the path of the file, so I'll sort in lexicographical order to preserve this behavior. Still plan on creating the CRDs before the other manifests.